### PR TITLE
Create DeleteOldZendeskTicketsJob

### PR DESCRIPTION
**Pull request by tvararu**
### Context

Create a job that will be called from a section in support to delete old Zendesk tickets.

### Changes proposed in this pull request

Review commit by commit.

### Guidance to review

I haven't really tested this end to end locally, because it's a fairly destructive job, but based on some testing in the `rails c` I don't see why this wouldn't work.

### Checklist

- [x] Attach to Trello card https://trello.com/c/fOknaOcV/401-zendesk-technical-analysis-of-ticket-deletion-6-months
- [x] Rebased main
- [x] Cleaned commit history
- [ ] Tested by running locally

### CODE REVIEW

File: app/jobs/delete_old_zendesk_tickets_job.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+class DeleteOldZendeskTicketsJob < ApplicationJob
+  def perform
+    tickets = ZendeskService.find_closed_tickets_from_6_months_ago
+    return if tickets.size.zero?
+
+    tickets.each { |ticket| DeletedZendeskTicket.new.from(ticket).save! }
+
+    ids = tickets.map(&:id)
+    ZendeskService.destroy_tickets!(ids)
```
[the remark this answers is not in the document]
Comment (tvararu) [reply app/jobs/delete_old_zendesk_tickets_job.rb:9]: "Accepts a comma-separated list of up to 100 ticket ids."

Good shout, I'll add some code to handle that.

File: app/jobs/delete_old_zendesk_tickets_job.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+class DeleteOldZendeskTicketsJob < ApplicationJob
+  def perform
+    tickets = ZendeskService.find_closed_tickets_from_6_months_ago
+    return if tickets.size.zero?
+
+    tickets.each { |ticket| DeletedZendeskTicket.new.from(ticket).save! }
```
[the remark this answers is not in the document]
Comment (tvararu) [reply app/jobs/delete_old_zendesk_tickets_job.rb:6]: Yup, I'll have a think.

File: app/jobs/delete_old_zendesk_tickets_job.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (tvararu) [reply app/jobs/delete_old_zendesk_tickets_job.rb:14]: On reflection, the search endpoint also only return 100 results; maybe recursively calling the job if it got exactly 100 results could work. But let's test it end to end first.

[the remark this answers is not in the document]
Comment (benilovj) [reply app/jobs/delete_old_zendesk_tickets_job.rb:19]: > On reflection, the search endpoint also only return 100 results; maybe recursively calling the job if it got exactly 100 results could work. But let's test it end to end first.

Just looking at the [search endpoint documentation](https://developer.zendesk.com/api-reference/ticketing/ticket-management/search/):

> The Search API returns up to 1,000 results per query, with a maximum of 100 results per page. See [Pagination](https://developer.zendesk.com/rest_api/docs/support/introduction#pagination). If you request a page past the limit (page=11 at 100 results per page), a 422 Insufficient Resource Error is returned.

Also the [results response](https://developer.zendesk.com/api-reference/ticketing/ticket-management/search/#json-format) has a `next_page` URL if there is a next page.

**Pull request merged**